Fix deadlock by closing pipe on exit

ck3d opened the pull request:
In 1 of 1000000 cases redo-ifchange gets blocked at a read call
in function try_procure (line 752 in redo.c) and the child process
ends in zombie state.
An additional usleep call before the read call causes this effect to
occur more often.

I do not know why, but closing the pipe during exit fixes this issue.

leahneukirchen commented:
pipe2 is not portable.

And closing the pipe doesn't help, because the children need that pipe for the thing to work?

Do you have a script to reproduce this?

leahneukirchen commented:
Potentially, two processes read from the pipe and one loses. Try making the pipe nonblocking?

```
fcntl(poolrd_fd, F_SETFL, O_NONBLOCK);
```

ck3d commented:
Thanks for your fast answer.
I could not reduce this problem into a simple example, but I would try harder if we do not find a solution for it.
Tomorrow I can give you feedback to your proposed fix.

leahneukirchen commented:
(add the code before the poll)

ck3d commented:
Thanks, your porposal fixed the deadlock!
I added your code snippet at [line 749](https://github.com/leahneukirchen/redo-c/blob/master/redo.c#L749) as you told me as replacement to my proposal. I ran a lot of tests and I am sure that this issue is now fixed for me.

My proposal is now obsolete and I will close this PR. If I should run some other tests, please let me know!

leahneukirchen commented:
Ok, I'll provide a cleaner patch for you to try.

leahneukirchen commented:
Try #4 please.

ck3d commented:
I performed same tests as before without any complaints, perfect.

Thanks for your great redo implementation and the fast and professional communication.
